aten/src/ATen/core/Allocator.h

```diff
@@ -133,3 +133,27 @@ struct CAFFE2_API InefficientStdFunctionContext {
 };
 
 } // namespace at
+
+namespace caffe2 {
+
+// using AllocatorArray = std::array<
+//     std::unique_ptr<at::Allocator>,
+//     static_cast<int>(at::DeviceType::COMPILE_TIME_MAX_DEVICE_TYPES)>;
+/** Set the allocator for DeviceType `t` and takes ownership of the allocator
```

Comment:
It seems a lot safer for this to just accept a `std::unique_ptr<at::Allocator>&&` directly, and have the call site make a unique pointer if they have a raw pointer they want to steal. At some point we should fix the `SetCPUAllocator` to something more safe (think about the API difference between `shared_ptr<V>(new V())` versus `make_shared<V>())`, but we can start by making the inner-most API have the safe types.

Reply:
BTW, this function is also dangerous, because it *invalidates* any other references to the allocator that was previously stored here. So any calls to this beyond the very beginning of this program are liable to cause big problems because we can't keep track if the allocator is still live or not.

When I designed the original allocator API, I wrote it under the assumptions that all allocators have static lifetime. So maybe this array should be written that way too: assume that every allocator has already taken care of its static lifetime somehow, and we're just going to take in a non-owning reference to keep track of it. Then it would even be safe to set the allocator mid-program (though you'd still have to make sure to synchronize all threads) because you wouldn't be invalidating any allocators.

Reply:
This is mostly because we have a `SetCPUAllocator` that will change the cpu allocator before,
for invalidation, I think we have the same problem before, so previously we have a `std::unique_ptr<CPUAllocator>` that owns the allocator and `SetCPUAllocator` will change the stored CPU allocator, I assume all the change is done during setup time and users won't be affected?